Ubuntu/History/historymanager.cpp

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright (C) 2021 UBports Foundation
+ *
+ * Authors:
+ *  Lionel Duboeuf <lduboeuf@ouvaton.org>
+ *
+ * This file is part of history-service.
+ *
+ * history-service is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; version 3.
+ *
+ * history-service is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include "historymanager.h"
+#include "manager.h"
+#include "eventview.h"
+#include <QDebug>
+#include <QTimer>
+
+
+HistoryManager::HistoryManager(QObject *parent) : QObject(parent), mError(OperationError::NO_ERROR), mDeletedCount(0), mPendingOperation(false)
+{
+}
+
+int HistoryManager::count() const
+{
+    return mCount;
+}
+
+int HistoryManager::deletedCount() const
+{
+    return mDeletedCount;
+}
+
+HistoryManager::OperationError HistoryManager::error() const
+{
+    return mError;
+}
+
+bool HistoryManager::removeAll(int eventType, const QString &from)
+{
+
+    if (mPendingOperation) {
+        setError(OperationError::OPERATION_ALREADY_PENDING);
+        qWarning() << "there is a pending operation, request cancelled";
+        return false;
+    }
+
+    QDateTime fromDate = QDateTime::fromString(from, Qt::ISODate);
+    History::EventType type = (History::EventType) eventType;
+
+    if (type == History::EventTypeNull|| !fromDate.isValid()) {
+        setError(OperationError::OPERATION_INVALID);
+        qWarning() << "invalid type or date, request cancelled";
+        return false;
+    }
+
+    History::Filter queryFilter(History::FieldTimestamp, QVariant(from), History::MatchLess);
+    History::Sort querySort(History::FieldTimestamp, Qt::DescendingOrder);
+
+    if (!queryFilter.isValid()) {
+        setError(OperationError::OPERATION_INVALID);
+        qWarning() << "filter invalid, operation cancelled";
+        return false;
+    }
+
+    if (!mView.isNull()) {
+        mView->disconnect(this);
+    }
+
+    setCount(History::Manager::instance()->eventsCount(type, queryFilter));
```

Comment:
Yes, we need to inform users the total count of events that needs to be removed. User will be notified of the number of removed events with the deletedCount property

Reply:
I see, thanks for the explanation. This new class is however a bit complicated to use, and does not really follow the way that asynchronous APIs are implemente in Qt/QML. I suggest you to refactor it in such a way that the asynchronous operation status is not exposed as direct members of this class, but keep it in the context of the asynchronous method. It would be nice if this class just exposed a single method to delete the items, and the result (number of deleted items or error) was delivered as a callback, like [QQuick's Item::grabToImage()](https://doc.qt.io/qt-5/qml-qtquick-item.html#grabToImage-method).

For an example how how to implement that, you can have a look at the [runHelper() method](https://gitlab.com/mardy/imaginario/-/blob/master/src/helper_model.h#L77) and [how it invokes a Javascript callback](https://gitlab.com/mardy/imaginario/-/blob/master/src/helper_model.cpp#L460-463).

It might sound complicated, but it will simplify your code a lot :-)

Reply:
ok, will look at it ^^

Reply:
but if you look at the client code, it is not so complicated :) : https://github.com/ubports/dialer-app/blob/0e1e2b3cc9893afb834deee4a1ca797af119c7c0/src/qml/HistoryPage/HistoryCleaner.qml